Add SrcOrder tag to compute_kernel_hw_startup (#22219, part 1)

ncvetkovicTT opened the pull request:
### PR Category
Feature

### Summary
First step of the matmul compute API cleanup (#22219). This PR only adds the `SrcOrder` tag to `compute_kernel_hw_startup`; the matmul API that uses it, and the kernel migration, follow in a separate PR so this one stays small and easy to review.

The end goal of #22219 is the same simple model for every op: call `compute_kernel_hw_startup` once at the start of the kernel, then call one `<op>_init` before the op. Matmul is the awkward case because it consumes its operands in the opposite order from every other op: in0 goes to SrcB and in1 goes to SrcA. `compute_kernel_hw_startup` programs the source registers (formats, tile and face sizes), so it has to know that order.

To handle it cleanly without leaking the swap into every kernel, this adds a small template tag:

- `SrcOrder::Regular` (the default) keeps the existing mapping (icb0 to SrcA, icb1 to SrcB), so **every current call site is unchanged**.
- `SrcOrder::Reverse` maps icb0 to SrcB and icb1 to SrcA, which is what matmul needs.

```cpp
compute_kernel_hw_startup<SrcOrder::Reverse>(in0, in1, out);  // matmul
matmul_init(in0, in1);
```

The tag is a template parameter, so the mapping is resolved at compile time, and both UNPACK and MATH `hw_configure` are programmed with the same ordering.

### Notes for reviewers
- One file: `compute_kernel_hw_startup.h`.
- Default is `SrcOrder::Regular`, so this is a no-op for all existing callers; nothing else in the tree needs to change for this PR.
- The matmul API (`matmul_init`, `matmul_block_init`), the deprecation of the old `mm_*` inits, and the migration of all matmul kernels to the new API will come in a follow-up PR (kept separate so the deprecation warnings land together with the kernel changes that remove them).

ncvetkovicTT commented:
### Pipeline results (head `3dbc333`, rebased on latest `main`)

Ran the four pipelines on the rebased PR head. This is an API only change to `matmul.h` and `compute_kernel_hw_startup.h`, and no kernels are migrated yet, so I do not expect any test impact. Results below, with the one red checked against `main`.

- **Sanity** — passes ([run 27073445716](https://github.com/tenstorrent/tt-metal/actions/runs/27073445716)).
- **Sanity (Runtime)** — passes ([run 27073446271](https://github.com/tenstorrent/tt-metal/actions/runs/27073446271)).
- **Nightly L2** — passes ([run 27073447496](https://github.com/tenstorrent/tt-metal/actions/runs/27073447496)).
- **Blackhole post-commit** — fails ([run 27073446887](https://github.com/tenstorrent/tt-metal/actions/runs/27073446887)), but the only red is `deepseek blitz op tests` (fast dispatch and slow dispatch). The same two jobs fail on the latest scheduled `main` run from just before this one ([run 27072898654](https://github.com/tenstorrent/tt-metal/actions/runs/27072898654)), so this is pre existing on `main` and not caused by this PR.

So everything that is green stays green, and the one failure also fails on `main`. No failure is introduced by this change.

ncvetkovicTT commented:
/codeowners ping

fvranicTT commented on tt_metal/hw/inc/api/compute/compute_kernel_hw_startup.h:
```diff
@@ -13,6 +13,28 @@
+// clang-format on
+enum class SrcOrder : uint8_t {
+    Regular = 0,  // icb0 -> SrcA, icb1 -> SrcB
+    Matmul = 1,   // icb0 -> SrcB, icb1 -> SrcA
```
A good generalized approach would be to steer away from impl details and call `SrcOrder::Matmul` a bit differently, so it stays general through time, even when some new implementation comes that might use that ordering of operands. Calling it maybe Reverse or somewhat different might be a safer choice.

ncvetkovicTT replied:
Accepted.

ncvetkovicTT replied:
Done in 6c74d08 — renamed the tag to `SrcOrder::Reverse` (and updated the docs and call examples). Kept matmul as the example of an op that needs it.

fvranicTT commented on tt_metal/hw/inc/api/compute/compute_kernel_hw_startup.h:
```diff
@@ -31,30 +59,39 @@ namespace ckernel {
+    // swapped. Both UNPACK and MATH hw_configure are programmed with the same (src_a_cb, src_b_cb)
+    // ordering so the unpacker tile descriptors and the math ALU format registers agree.
+    constexpr bool swap = (src_order == SrcOrder::Matmul);
+    const uint32_t src_a_cb = swap ? icb1 : icb0;
```
This could also be a compile time construct, if we use the src_order for comparison inside if constexpr.

ncvetkovicTT replied:
Done in 6c74d08 — the operand mapping now uses `if constexpr (src_order == SrcOrder::Reverse)` so the swap is resolved at compile time.

fvranicTT commented on tt_metal/hw/inc/api/compute/matmul.h:
```diff
@@ -72,52 +72,6 @@ ALWI void matmul_block_math_dynamic_throttle(
- * | transpose      | The transpose flag for performing transpose operation on B    | uint32_t | Any positive value will indicate transpose is set  | False    |
- */
-// clang-format on
-ALWI void mm_init(
```
These functions were never deprecated (applies to other functions too). Are we allowed to just remove them without deprecation step? I’d say we would need to live with them for another month and add deprecation annotation designating them for removal.

fvranicTT replied:
Ah I now see you moved them to the end of the file. Sorry I missed that initially.

ncvetkovicTT replied:
No worries, I was wondering if I should make a note to reviewers because I missed that too when looking at the changes..

ncvetkovicTT commented:
@fvranicTT @skrsmanovicTT PTAL. Stefan, you filed an issue about this a while back, I linked it in the PR.

ncvetkovicTT commented:
### Nightly L2 re-run with all subgroups (head `de28a37`)

My first Nightly L2 dispatch had every test subgroup skipped (a `workflow_dispatch` run defaults all groups to off), so it was not a real run. This time I enabled all subgroups. The matmul related groups all pass, and every red also fails on `main`.

**Passing — the groups this change could plausibly affect:**
- `ttnn nightly matmul` — passes on WH (N150, N300) and BH (P100a, P150b)
- `ttnn nightly reduction`, `pool`, `eltwise` (groups 1-4), and `sdpa` nightly + stress — all pass on WH and BH

**Red, and all pre-existing on `main`:** `ttnn nightly conv`, `compute-fused validation`, `sdxl`, `experimental`, `moreh`, `data_movement`, `tt-train` python unit tests, and `cpp ttnn tensor accessor`. The latest scheduled `main` run from this morning ([run 27085942812](https://github.com/tenstorrent/tt-metal/actions/runs/27085942812)) fails the same set. None of these exercise matmul init.

**One difference from `main`:** `didt tests (WH N300)` failed here but passes on `main`. didt is a power/current measurement test that is runner sensitive — in this same run didt passed on WH N150 and on both BH cards, and it passes on `main`, so this looks like runner flakiness rather than a code issue. This PR only edits two matmul headers and migrates no kernels.

**Sanity** — passes on `de28a37` ([run 27082987246](https://github.com/tenstorrent/tt-metal/actions/runs/27082987246)). I am re-running it on the latest head `6c74d08` (the `SrcOrder::Reverse` rename plus `if constexpr` refactor) so the final commit has its own green build.

Net: matmul passes on all cards, and no failure is introduced by this PR.

amahmudTT commented:
You mentioned the reverse behavior as ino going to srcB, isnt it the old behavior ?  Asking because you have mentioned the default is the old behavior and reverse is the behavior in which in-0 goes to srcB

amahmudTT commented on tt_metal/hw/inc/api/compute/compute_kernel_hw_startup.h:
```diff
@@ -31,30 +59,43 @@ namespace ckernel {
+    // the operands swapped. The mapping is resolved at compile time. Both UNPACK and MATH hw_configure
+    // are programmed with the same (src_a_cb, src_b_cb) ordering so the unpacker tile descriptors and the
+    // math ALU format registers agree.
+    uint32_t src_a_cb = icb0;
```
Nit : you could do this way without assigning the values and then changing them.

uint32_t src_a_cb =  SrcOrder::Reverse ? icb1:icb0;

ncvetkovicTT replied:
Yeah idk why it changed it in this ugly way, I'll change it back to what you suggest.

ncvetkovicTT replied:
Done in 9d94200 — switched to a `constexpr bool reverse` + ternary, so no more default-then-overwrite. Kept it constexpr so the mapping is still resolved at compile time.

ncvetkovicTT commented:
> You mentioned the reverse behavior as ino going to srcB, isnt it the old behavior ? Asking because you have mentioned the default is the old behavior and reverse is the behavior in which in-0 goes to srcB

Regular: in0 -> SrcA, in1 -> SrcB (all ops except matmul are like this)
Reverse: in0 -> SrcB, in1 -> SrcA (currently only matmul uses this)

ncvetkovicTT commented on tt_metal/hw/inc/api/compute/matmul.h:
```diff
@@ -161,16 +118,25 @@ ALWI void matmul_tiles(
+ *
+ * It is safe to call this multiple times to switch back to matmul mode after another operation, as long
+ * as the operand data formats still match those configured at startup. If the operand data formats
+ * changed, use matmul_init_reconfig_data_format_srca() / matmul_init_reconfig_data_format() instead.
```
Addressed in de28a37 — added the missing helpers, so the docs now point to real functions: matmul_init_reconfig_data_format_srca / _srcb / matmul_init_reconfig_data_format (and the same three for the block variants).

ndivnicTT commented on tt_metal/hw/inc/api/compute/matmul.h:
```diff
@@ -209,69 +178,72 @@ ALWI void mm_init_short(
  */
 // clang-format on
-ALWI void mm_init_short_with_dt(
+ALWI void matmul_init_reconfig_data_format_srca(
```
I am not sure that renaming is needed in this PR, especially since this function will be erased in near future.
I like the new name, just think it is not needed

ncvetkovicTT committed:
Add SrcOrder tag to compute_kernel_hw_startup

Adds an enum class SrcOrder { Regular, Reverse } and makes compute_kernel_hw_startup
a template parameterized on it. Default SrcOrder::Regular keeps the existing
icb0->SrcA, icb1->SrcB mapping, so all current call sites are unchanged.
SrcOrder::Reverse maps icb0->SrcB, icb1->SrcA, which matmul needs (in0->SrcB,
in1->SrcA); it is resolved at compile time and programs both UNPACK and MATH
hw_configure with the same ordering. The matmul API that uses this tag follows in
a separate PR.

ndivnicTT reviewed (approved):
LGTM!

ncvetkovicTT commented:
/codeowners ping
